Support installing requirements from inline script metadata (PEP 723)

SnoopJ opened the pull request:
This changeset adds support for installing requirements declared in a script using inline metadata as defined by [PEP 723](https://peps.python.org/pep-0723/)

~I'm sure this still needs changes, but I'm at the point where I'm ready for a maintainer to point out to me how much I've missed 😅~

Closes #12891

SnoopJ committed:
Add --script to install command

SnoopJ committed:
Add failing test for --script

SnoopJ committed:
Default --script to None

SnoopJ committed:
Add minimum implementation of parsing requirements from inline metadata

SnoopJ committed:
Issue an error if --script is given multiple times

This also lays the groundwork for potentially allowing the argument to
be given multiple times, but I don't yet want to think about the
additional complexity of multiple possibly-contradictory
`requires-python` specifiers.

SnoopJ committed:
Add scripts() to download, wheel subcommands

SnoopJ committed:
Test that --script can only be given once

SnoopJ committed:
Remove TODO (I think the answer is 'no')

SnoopJ committed:
Add failing test for incompatible requires-python

SnoopJ committed:
Correct type annotation of PEP 723 helper

SnoopJ committed:
Remove PEP 723 requirements helper in favor of direct access

SnoopJ committed:
Check requires-python specified in script metadata

SnoopJ committed:
Appease the linters

SnoopJ committed:
Write return annotation correctly

SnoopJ committed:
Add NEWS fragment

SnoopJ commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -268,11 +277,37 @@ def get_requirements(
                 )
                 requirements.append(req_to_add)
 
+        if options.scripts:
+            if len(options.scripts) > 1:
+                raise CommandError("--script can only be given once")
+
+            script = options.scripts[0]
+            script_metadata = pep723_metadata(script)
+
+            script_requires_python = script_metadata.get("requires-python", "")
+
+            if script_requires_python and not options.ignore_requires_python:
+                target_python = make_target_python(options)
+
+                if not check_requires_python(
+                    requires_python=script_requires_python,
+                    version_info=target_python.py_version_info,
+                ):
+                    raise UnsupportedPythonVersion(
+                        f"Script {script!r} requires a different Python: "
+                        f"{target_python.py_version} not in {script_requires_python!r}"
+                    )
+
+            for req in script_metadata.get("dependencies", []):
+                requirements.append(
+                    InstallRequirement(Requirement(req), comes_from=None)
+                )
```
It feels ungraceful to put all of this logic here but I'm not really sure how I would plumb it into a separate function. Suggestions welcome

SnoopJ commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -268,11 +277,37 @@ def get_requirements(
+                        f"{target_python.py_version} not in {script_requires_python!r}"
+                    )
+
+            for req in script_metadata.get("dependencies", []):
+                requirements.append(
```
`ruff` is upset that I wrote a loop with `.append()` here (`PERF401`) but I don't know if a comprehension would really be justified here? I can maybe see the case for `requirements.extend(...)` but it still seems like it'd be harder to read to me. Looking for feedback.

SnoopJ committed:
Change --script to --requirements-from-script

SnoopJ committed:
Replace --script usage in dedicated PEP 723 tests, fix INITools naming issue

SnoopJ committed:
Add --requirements-from-scripts to lock command

SnoopJ committed:
Add extra layer of backslash escaping to match test stderr on Windows

SnoopJ committed:
Remove obsolete TODO

SnoopJ committed:
Change typing.Dict[] to dict[] to appease ruff-check

SnoopJ committed:
Adjust test requires-python to 'not this version'

Note that a trailing .* is necessary to make the resulting SpecifierSet
broad enough to match any minor version.

Co-authored-by: Paul Moore <p.f.moore@gmail.com>

SnoopJ committed:
Simplify test of expected error message

Co-authored-by: Paul Moore <p.f.moore@gmail.com>

SnoopJ committed:
Merge branch 'main' into feature/gh12891-inline-metadata

SnoopJ committed:
Fix syntax error

That's what I get for trying to take the web UI seriously.

ichard26 commented on tests/functional/test_install_script.py:
```diff
@@ -0,0 +1,96 @@
+    result.did_create(script.site_packages / "initools")
+    assert result.files_created[script.site_packages / other_lib_name].dir
+    fn = f"{other_lib_name}-{other_lib_version}.dist-info"
+    assert result.files_created[script.site_packages / fn].dir
```
The `result.assert_installed()` can be used instead:

- `result.assert_installed("package")` to simply do a name check
- `result.assert_installed(package="1.0.0")` to do a name and version check

ichard26 reviewed (changes requested):
First pass review. I'll have to think about the option naming some more, but overall, I'm supportive of the feature. While I do agree this is a suboptimal use of inline script metadata, if pip is meant to be a standards package installer, it should be able to install requirements declared in any standardized format.

SnoopJ committed:
Remove comment put in place for sake of review

SnoopJ committed:
Use 'name' in error

SnoopJ committed:
Use CommandError for PEP 723 parsing failures

SnoopJ committed:
Avoid network use by tests, simplify assertions

Co-authored-by: Richard Si <sichard26@gmail.com>

SnoopJ committed:
Refine requirements from script metadata

Co-authored-by: Richard Si <sichard26@gmail.com>

SnoopJ commented on src/pip/_internal/cli/req_command.py:
```diff
@@ -305,6 +314,32 @@ def get_requirements(
+                raise CommandError("--requirements-from-script can only be given once")
+
+            script = options.requirements_from_scripts[0]
+            script_metadata = pep723_metadata(script)
```
Good call.

Changed these `ValueError` to a PEP 723-specific exception that is translated to `CommandError` at the site of use (it doesn't seem appropriate to raise it directly), in the hopes that no _implicit_ `ValueError` would escape (since those would be bugs).

SnoopJ commented on tests/functional/test_install_script.py:
```diff
@@ -0,0 +1,96 @@
+    ), ("multiple script did not fail as expected -- " + result.stderr)
+
+
+@pytest.mark.network
```
Resolved in 003a443

SnoopJ commented on tests/functional/test_install_script.py:
```diff
@@ -0,0 +1,96 @@
+            """
+        )
+    )
+    result = script.pip("install", "--requirements-from-script", script_path)
```
Thanks for pointing me to the right mechanism for local-only installation. Resolved in 003a443

SnoopJ committed:
Remove unused import

ichard26 commented on news/12891.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Support installing dependencies declared with inline script metadata
+(PEP 723).
```
```suggestion
Support installing dependencies declared with inline script metadata
(:pep:`723`) with `--requirements-from-script`.
```

ichard26 commented on src/pip/_internal/metadata/pep723.py:
```diff
@@ -0,0 +1,35 @@
+            line[2:] if line.startswith("# ") else line[1:]
+            for line in matches[0].group("content").splitlines(keepends=True)
+        )
+        return tomllib.loads(content)
```
This can also raise an exception if the TOML is invalid.

SnoopJ replied:
Good point, resolved in [f42ae07](https://github.com/pypa/pip/pull/13052/commits/f42ae0726a40afb49104da1624ac288d754da28e) and tested in [f459c7a](https://github.com/pypa/pip/pull/13052/commits/f459c7a4324cede34d96f8016ba2c0284430fb07) along with the multiple-blocks failure mode.

ichard26 reviewed (commented):
The tests are much nicer, indeed!

Other than my remaining inline feedback, my last concern is the inclusion of the `-s` short option. I don't have particularly well-defined reasons for why I'm hesitant to include it other than that I don't want `pip install -s` to become popular, especially since it's not the most clear with what it does.

pfmoore commented:
> Other than my remaining inline feedback, my last concern is the inclusion of the -s short option.

I agree. I think we can reasonably omit the `-s` option for now, and include it later *if* there's sufficient demand for it. It's convenient, but I agree that it's not obvious.

SnoopJ committed:
Remove short option for PEP 723 requirements

SnoopJ committed:
Move PEP 723 utilities to req

SnoopJ committed:
Add tests for failure modes

Covers failures due to repeated 'script' metadata blocks and invalid
TOML metadata.

SnoopJ commented on src/pip/_internal/metadata/pep723.py:
```diff
@@ -0,0 +1,35 @@
+        filter(lambda m: m.group("type") == name, re.finditer(REGEX, script))
+    )
+    if len(matches) > 1:
+        raise PEP723Exception(f"Multiple {name} blocks found")
```
Resolved in [f42ae07](https://github.com/pypa/pip/pull/13052/commits/f42ae0726a40afb49104da1624ac288d754da28e)

SnoopJ commented on src/pip/_internal/req/pep723.py:
Resolved in [f42ae07](https://github.com/pypa/pip/pull/13052/commits/f42ae0726a40afb49104da1624ac288d754da28e)

SnoopJ committed:
Fixup reST in NEWS entry

Co-authored-by: Richard Si <sichard26@gmail.com>

SnoopJ commented:
Thanks, made those changes.

I also added two new tests for the failure modes, which turned out to be informative because it turns out that the [reference implementation](https://peps.python.org/pep-0723/#reference-implementation) has surprising match behavior for multiple inline metadata blocks _unless_ there is a line that **does not** start with `\s*#` separating the two. If they aren't separated like that, it's all collated into one match, no matter how many `/// TYPE` blocks there are inside of that match. This probably makes more sense with a code example:

<details><summary>click for code</summary>

```
import re

script_A = """
# /// script
# data (1)
# ///
#
# /// script
# data (2)
# ///
"""

script_B = """
# /// script
# data (1)
# ///

# /// script
# data (2)
# ///
"""

# These lines adapted from PEP 723's reference parser:
# https://peps.python.org/pep-0723/#reference-implementation

REGEX = r"(?m)^# /// (?P<type>[a-zA-Z0-9-]+)$\s(?P<content>(^#(| .*)$\s)+)^# ///$"
name = "script"
matches_A = list(
    filter(lambda m: m.group("type") == name, re.finditer(REGEX, script_A))
)
matches_B = list(
    filter(lambda m: m.group("type") == name, re.finditer(REGEX, script_B))
)

# output:
# 1
# 2
print(len(matches_A))
print(len(matches_B))
```

</details>

The upshot is that the parser used here would still fail when calling `loads()` on the match (because the interior `///` lines are invalid TOML), which meets PEP 723's `MUST` error requirement. But I thought I'd call it out in case it's cause enough to modify the reference implementation to detail with this edge case.

EDIT: changing the pattern's last `+` to `+?` to make the match non-greedy seems to fix this edge case, so maybe the change is simple.

ichard26 reviewed (approved):
> Edit: looks like I may have caused a signoff problem with marking https://github.com/pypa/pip/commit/003a4430b78ea36e14a9f2c654651cc4933f0ea5 as Co-authored-by, let me know if the history needs rewriting to deal with that.

I'll squash merge this PR, so it won't matter. Don't worry about it!

> EDIT: changing the pattern's last + to +? to make the match non-greedy seems to fix this edge case, so maybe the change is simple.

Hmm, as much as I would like to ensure error message accuracy, I am wary of deviating from the reference implementation. We may just be swapping one edge case for another, but at least we can hand-wave the current one away as "it's what the reference implementation does." I wonder if this has been flagged to the PEP 723 authors yet.

Thanks a lot for bearing with my nit-picky reviews! I'll wait for @notatallshaw to review as he did self-request a review, but from my end, this looks great!

SnoopJ commented:
I agree about sticking to the reference implementation, especially as the edge-case is already so small since there's only one `TYPE` currently recognized by the official standards.

I've filed an [issue](https://github.com/pypa/packaging.python.org/issues/1960) with the PyPA specs for broader discussion.

ichard26 committed:
Fix backtick syntax in news/12891.feature.rst

notatallshaw reviewed (approved):
Looks good to me.
